odoo/tools/convert.py

```diff
@@ -86,30 +75,75 @@ def _fix_multiple_roots(node):
+# ideally would have used decorator.decorator, but the type inference doesn't
+# work
+def with_data(callback):
+    @functools.wraps(callback)
+    def _data(self, node, _env):
+        data = node.text.encode()
+        if node.get('file'):
+            with file_open(node.get('file'), 'rb') as f:
+                data = f.read()
+        return callback(self, data or b'')
+    return _data
+
+def with_text(callback):
+    @functools.wraps(callback)
+    def _text(self, node, _env):
+        text = node.text
+        if node.get('file'):
+            with file_open(node.get('file'), 'rb') as f:
+                text = pycompat.to_text(f.read())
+        return callback(self, text or '')
+    return _text
+
+def _eval_xml(converter, node, env):
+    return Evaluator(converter).evaluate(node, env)
+
+class Evaluator:
+    def __init__(self, converter):
+        self.converter = converter
+
+    @property
+    def idref(self):
+        return self.converter.idref
+    @property
+    def id_get(self):
+        return self.converter.id_get
+    @property
+    def module(self):
+        return self.converter.module
+
+    def evaluate(self, node, env):
+        # <function>
+        #   <function/>
+        # </function>
+        if node.tag == 'function':
+            return self.converter._tag_function(node)
+
+        from odoo import models # circular dependencies ho!
+        assert node.tag in ('field','value')
+
         f_model = node.get('model')
-        if node.get('search'):
-            f_search = node.get("search")
-            f_use = node.get("use",'id')
-            f_name = node.get("name")
-            idref2 = {}
-            if f_search:
-                idref2 = _get_idref(self, env, f_model, self.idref)
+
+        # this is only for <value search=""> as <field> is handled separately
+        # in <record>'s callback
+        f_search = node.get("search")
+        if f_search:
+            idref2 = _get_idref(self, env, f_model, self.idref)
             q = safe_eval(f_search, idref2)
-            ids = env[f_model].search(q).ids
-            if f_use != 'id':
-                ids = [x[f_use] for x in env[f_model].browse(ids).read([f_use])]
-            _fields = env[f_model]._fields
-            if (f_name in _fields) and _fields[f_name].type == 'many2many':
-                return ids
-            f_val = False
-            if len(ids):
-                f_val = ids[0]
-                if isinstance(f_val, tuple):
-                    f_val = f_val[0]
-            return f_val
+
+            # can't limit=1 here or use on a relational field might get 0
+            # records
+            results = env[f_model].search(q)
+            f_use = node.get('use')
+            if f_use is not None:
+                results = results.mapped(f_use)
+            first_result = results[:1]
+            if isinstance(results, models.Model):
+                return first_result.ids or False
+            return first_result or False
```

Comment:
These return lists, whereas the old version would return just one value (unless we were deref'ing an m2m but that sounds weird)